src/helpers.cpp

```diff
@@ -151,4 +151,182 @@ Vec3 Triangulate2(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+
+  // Initalize the Rotation and Translation Matricies
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc1(0), xc1(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc2(0), xc2(1), 1.0};
+  f1.normalize();  
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  Vec3 m0_hat = m0 / m0.norm();
+  Vec3 m1_hat = m1 / m1.norm();
+
+  Eigen::Matrix<double, 3, 2> A;
+  A.row(0) << m0_hat(0), m1_hat(0);
+  A.row(1) << m0_hat(1), m1_hat(1);
+  A.row(2) << m0_hat(2), m1_hat(2);
+
+  Vec3 t21_hat = t21 / t21.norm();
+  Mat3 I = Eigen::Matrix3d::Identity();
+
+  Eigen::Matrix<double, 2, 3> B;
+  B = A.transpose() * (I - t21_hat * t21_hat.transpose());
+
+  Eigen::JacobiSVD<Eigen::Matrix<double, 2, 3>> svd(B, Eigen::ComputeFullV);
+  Mat3 V = svd.matrixV();
+  Vec3 n_prime_hat = V.col(1);
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 z = m1_prime.cross(m0_prime);
+
+  Vec3 x = ((z.dot(t21.cross(m0_prime))) / pow(z.norm(),2)) * m1_prime;
+
+  return x;
+}
+
+Vec3 Triangulate5(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+
+  // Initalize the Rotation and Translation Matricies
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc1(0), xc1(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc2(0), xc2(1), 1.0};
+  f1.normalize();  
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  Vec3 m0_hat = m0 / m0.norm();
+  Vec3 m1_hat = m1 / m1.norm();
+
+  Vec3 n_a = (m0_hat + m1_hat).cross(t21);
+  Vec3 n_b = (m0_hat - m1_hat).cross(t21);
+
+  Vec3 n_prime_hat = n_a.norm() >= n_b.norm() ? n_a : n_b;
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 z = m1_prime.cross(m0_prime);
+
+  // check_cheirality(z, t21, m1_prime, m0_prime);
+  // check_angular_reprojection(m0, m0_prime, m1, m1_prime);
+  // check_parallax(m0_prime, m1_prime);
+
+  Vec3 x = ((z.dot(t21.cross(m0_prime))) / pow(z.norm(), 2)) * m1_prime;
+
+  return x;
+}
+
+
+void check_cheirality(const Vec3 &z, const Vec3 &t, const Vec3 &f1_prime, const Vec3 &Rf0_prime)
+{
+
+  float lambda0 = z.dot(t.cross(f1_prime)) / pow(z.norm(), 2);
+  float lambda1 = z.dot(t.cross(Rf0_prime)) / pow(z.norm(), 2);
+
+  if(lambda0 <= 0 || lambda1 <= 0)
+  {
+    std::cout << "[ERROR] cheirality error in triangulation. lambda0=" << lambda0 << ", lamba1=" << lambda1 << std::endl;
+    exit(1);
+  }
+
+  return;
+}
+
+
+void check_angular_reprojection(const Vec3 &Rf0, const Vec3 &Rf0_prime, const Vec3 &f1, const Vec3 &f1_prime)
+{
+
+  float theta0 = acos(Rf0.dot(Rf0_prime) / (Rf0.norm() * Rf0_prime.norm()));
+  float theta1 = acos(f1.dot(f1_prime) / (f1.norm() * f1_prime.norm()));
+
+  float max_theta = std::max(theta0, theta1);
+
+  if(max_theta > 0.01)
```

Comment:
`max_theta`, `beta`, and any other threshold parameters should be runtime options in the `.cfg` files and not hard-coded.